Add developer guideline for validation (#12811)
gardener/gardener: docs/development/validation-guidelines.md

### Validation for CustomResourceDefinitions

The CustomResourceDefinitions (CRDs) allow validation to be specified using [OpenAPI v3.0 validation](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation).
It is convenient to perform relatively simple validation checks.
For example, validate field is immutable or validate a slice contains at least one item.
However, more advanced validation is hard to express via these means and is performed by validating webhooks.

#### `extensions.gardener.cloud` API group

The extensibility story is based on the CRDs from the `extensions.gardener.cloud` API group.
They are validated by the `gardener-resource-manager`'s validating webhook. See [Extension Resource Validation](../concepts/resource-manager.md#extension-resource-validation).
The validation functions are defined in the [`github.com/gardener/gardener/pkg/apis/extensions/validation`](https://pkg.go.dev/github.com/gardener/gardener/pkg/apis/extensions/validation) package.

#### `operator.gardener.cloud` API group

The [gardener-operator](../concepts/operator.md) is responsible for the management of the garden cluster environment.

> It does this using the `Garden` and `Extension` CRDs from the `operator.gardener.cloud` API group.

Reviewer: ```suggestion
It does this by using the `Garden` and `Extension` CRDs from the `operator.gardener.cloud` API group.
```

Author: I pushed 08972c934dbd878169b4d7b4125244d4f85928b1.

  - The `componentbaseconfigv1alpha1.ClientConnectionConfiguration` type contains details for constructing a client against a Kubernetes cluster.
  - Use this validation function to validate fields of type `componentbaseconfigv1alpha1.ClientConnectionConfiguration` in component configuration APIs. See [example usage](https://github.com/gardener/gardener/blob/874791cb97e96d622c994c6cc8ef4091fc9e9102/pkg/controllermanager/apis/config/v1alpha1/validation/validation.go#L22).
- [`validationutils.ValidateLeaderElectionConfiguration`](https://pkg.go.dev/github.com/gardener/gardener/pkg/utils/validation#ValidateLeaderElectionConfiguration)
  - The `componentbaseconfigv1alpha1.LeaderElectionConfiguration` type defines the configuration of leader election clients for components that can run with leader election enabled.
  - Use this validation function to validate fields of type `componentbaseconfigv1alpha1.LeaderElectionConfiguration` in component configuration APIs. See [example usage](https://github.com/gardener/gardener/blob/874791cb97e96d622c994c6cc8ef4091fc9e9102/pkg/controllermanager/apis/config/v1alpha1/validation/validation.go#L23).

> ## General Guidelines

Reviewer: I would move this section in the beginning as the document will be most probably re-used for other components, it would be nice to start with the generic guideline and then go into g/g specifics.

Reviewer 2: I agree

Author: I added a separate document for validation guidelines for extensions. I hope this solves your comment.

If parts are unclear or new learnings arise, this guide should be adapted accordingly.

## The Importance of Validation

Validation is a critical part of maintaining the reliability, security, and overall quality of the Gardener project. It is important for several key reasons:
- **Security**
  - Validation protects against malicious inputs (like code injection), bypassing intended constraints, exploiting assumptions in the code and other attacks. The unexpected or harmful input is rejected before it can do damage.
- **API Consistency and Robustness**
  - Validating the resource ensures that incorrect or ambiguous data is caught early, preventing downstream controllers from acting on invalid input. Ensuring well-formed and valid API resource specifications is fundamental for successful lifecycle operations on these resources (for example successful Shoot cluster creation, consistent Shoot cluster state).
- **System Stability**
  - Invalid configurations passed unchecked into the system can result in unpredictable behavior or runtime errors. Early validation helps maintain system stability and predictability.
- **Clear and Immediate Feedback**

>   - By catching errors in the admission phase of a request, the API users get fast, actionable feedback. The API users don't have to wait an internal system operation to fail after minutes to receive a feedback that the provided input is invalid.

Reviewer 3: ```suggestion
  - By catching errors in the admission phase of a request, the API users get fast, actionable feedback. The API users don't have to wait an internal system operation to fail after several minutes to receive feedback that the provided input is invalid.
```

Nit

## Validation in Gardener

> Kubernetes allows defining custom resources using extension API server and CustomResourceDefinitions.

Reviewer 3: ```suggestion
Kubernetes allows defining custom resources using [extension API server](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/) and [CustomResourceDefinitions](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/).
```

Gardener is using both of the approaches.
The sections below [Validation for API resources](#validation-for-api-resources) and [Validation for CustomResourceDefinitions](#validation-for-customresourcedefinitions) describe how validation is handled for these two approaches.

> ### Validation for API resources

Reviewer 3: ```suggestion
### Validation for API Resources
```

Nit, to be consistent with other headings.

[Gardener API server](../concepts/apiserver.md) is an extension API server.

> An extension API server allows validation of a resource to be performed in the storage layer or in admission plugin.

Reviewer 3: ```suggestion
An extension API server allows validation of a resource to be performed in the storage layer or in admission plugins.
```

An admission plugin can accept a configuration. The configuration is specified in API server admission configuration. See [Gardener API server admission configuration example](../../example/20-admissionconfig.yaml).
For reference, check out Gardener API server admission plugins that accepts configuration:
- [`ShootTolerationRestriction`](https://pkg.go.dev/github.com/gardener/gardener/plugin/pkg/shoot/tolerationrestriction/apis/shoottolerationrestriction)
- [`ShootDNSRewriting`](https://pkg.go.dev/github.com/gardener/gardener/plugin/pkg/shoot/dnsrewriting)
- [`ShootResourceReservation`](https://pkg.go.dev/github.com/gardener/gardener/plugin/pkg/shoot/resourcereservation)

#### Writing Validation Code in API Server Admission Plugins

- Write the validation code for a field in an API server admission plugin if the validation logic requires checking the presence of another resource or its specification. Otherwise, write the validation code in the storage layer.

>   - Example: Shoot's `.spec.region` field cannot be validated only in the storage layer. The validation logic for this field checks if the Shoot's CloudProfile supports the region in question. This is an example of validation logic that requires checking the specification of a referenced resource. This validation must be performed in admission plugin. It is  performed in the [`ShootValidator` admission plugin](../concepts/apiserver-admission-plugins.md#shootvalidator) ([Shoot `.spec.region` validation](https://github.com/gardener/gardener/blob/f6fb7e2ca019fdd2a09c0a5da6475bf5d6bd2430/plugin/pkg/shoot/validator/admission.go#L1617-L1637)).

Reviewer 3: ```suggestion
  - Example: Shoot's `.spec.region` field cannot be validated only in the storage layer. The validation logic for this field checks if the Shoot's CloudProfile supports the region in question, see [`ShootValidator` admission plugin](../concepts/apiserver-admission-plugins.md#shootvalidator) ([Shoot `.spec.region` validation](https://github.com/gardener/gardener/blob/f6fb7e2ca019fdd2a09c0a5da6475bf5d6bd2430/plugin/pkg/shoot/validator/admission.go#L1617-L1637)). This is an example of validation logic that requires cross checking against another resource's specification. Hence, validation must be performed in an admission plugin.
```

#### `extensions.gardener.cloud` API group

The extensibility story is based on the CRDs from the `extensions.gardener.cloud` API group.
They are validated by the `gardener-resource-manager`'s validating webhook. See [Extension Resource Validation](../concepts/resource-manager.md#extension-resource-validation).
The validation functions are defined in the [`github.com/gardener/gardener/pkg/apis/extensions/validation`](https://pkg.go.dev/github.com/gardener/gardener/pkg/apis/extensions/validation) package.

#### `operator.gardener.cloud` API group

The [gardener-operator](../concepts/operator.md) is responsible for the management of the garden cluster environment.
It does this by using the `Garden` and `Extension` CRDs from the `operator.gardener.cloud` API group.
The CRDs are validated by the `gardener-operator`'s validating webhook. See [`gardener-operator` validation section](../concepts/operator.md#validation).
The validation functions are defined in the [`github.com/gardener/gardener/pkg/apis/operator/v1alpha1/validation`](https://pkg.go.dev/github.com/gardener/gardener/pkg/apis/operator/v1alpha1/validation) package.

### Validation of Component Configurations

> Most Gardener components have a component configuration that follows similar principles to the Gardener API.

Reviewer 3: ```suggestion
Most Gardener components have a component configuration that follows similar validation principles to those of the Gardener API.
```

### Utility functions from Gardener

Gardener offers several packages that contain validation utilities:
- [`github.com/gardener/gardener/pkg/utils/validation/cidr`](https://pkg.go.dev/github.com/gardener/gardener/pkg/utils/validation/cidr), aliased as `cidrvalidation`
  - The package provides validation utilities for CIDR ranges - validate if CIDR range is valid, if it overlaps with other CIDR ranges, etc.
  - The [`cidrvalidation.CIDR`](https://pkg.go.dev/github.com/gardener/gardener/pkg/utils/validation/cidr#CIDR) interface contains several utility functions for working with CIDR ranges. An implementation of the interface can be instantiated using the [`cidrvalidation.NewCIDR`](https://pkg.go.dev/github.com/gardener/gardener/pkg/utils/validation/cidr#NewCIDR) function.
  - The package also contains validation utilities for checking if Shoot networks intersect with Seed networks or if Shoot networks intersect between each other.
- [`github.com/gardener/gardener/pkg/utils/validation/kubernetes/core`](https://pkg.go.dev/github.com/gardener/gardener/pkg/utils/validation/kubernetes/core), aliased as `kubernetescorevalidation`

>   - The package contains copy of several utility functions from [k8s.io/kubernetes/apis/core/validation](https://pkg.go.dev/k8s.io/kubernetes/apis/core/validation). This is done to prevent importing the `k8s.io/kubernetes` dependency in Gardener.

Reviewer 4: ```suggestion
  - The package contains copy of several utility functions from [k8s.io/kubernetes/apis/core/validation](https://pkg.go.dev/k8s.io/kubernetes/pkg/apis/core/validation). This is done to prevent importing the `k8s.io/kubernetes` dependency in Gardener.

```

- [`github.com/gardener/gardener/pkg/utils/validation`](https://pkg.go.dev/github.com/gardener/gardener/pkg/utils/validation), aliased as `validationutils`
  - The package provides validation utilities for component configuration APIs.

Frequently used validation functions:
- [`kubernetescorevalidation.ValidateResourceQuantityValue`](https://pkg.go.dev/github.com/gardener/gardener/pkg/utils/validation/kubernetes/core#ValidateResourceQuantityValue)

>   - Use this function to validate fields of type `resource.Quantity`. See [example usage](https://github.com/gardener/gardener/blob/f6fb7e2ca019fdd2a09c0a5da6475bf5d6bd2430/pkg/apis/core/validation/shoot.go#L2066-L2068).

Reviewer 4: The link does not show the use of `ValidateResourceQuantityValue`.

  - The `componentbaseconfigv1alpha1.ClientConnectionConfiguration` type contains details for constructing a client against a Kubernetes cluster.
  - Use this function to validate fields of type `componentbaseconfigv1alpha1.ClientConnectionConfiguration` in component configuration APIs. See [example usage](https://github.com/gardener/gardener/blob/f6fb7e2ca019fdd2a09c0a5da6475bf5d6bd2430/pkg/controllermanager/apis/config/v1alpha1/validation/validation.go#L22).
- [`validationutils.ValidateLeaderElectionConfiguration`](https://pkg.go.dev/github.com/gardener/gardener/pkg/utils/validation#ValidateLeaderElectionConfiguration)
  - The `componentbaseconfigv1alpha1.LeaderElectionConfiguration` type defines the configuration of leader election clients for components that can run with leader election enabled.
  - Use this function to validate fields of type `componentbaseconfigv1alpha1.LeaderElectionConfiguration` in component configuration APIs. See [example usage](https://github.com/gardener/gardener/blob/f6fb7e2ca019fdd2a09c0a5da6475bf5d6bd2430/pkg/controllermanager/apis/config/v1alpha1/validation/validation.go#L23).

## Field-level Validation Errors

> The [`k8s.io/apimachinery/pkg/util/validation/field`](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/validation/field) package contains the predefined field-level errors.

Reviewer 3: ```suggestion
The [`k8s.io/apimachinery/pkg/util/validation/field`](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/validation/field) package contains the predefined field-level errors. These are commonly used by the validation logic to return detailed error information about specific fields in API objects.
```

  - A good practice for a webhook is to filter only for the resources it needs to validate by using an [`objectSelector`](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-objectselector).
  - Ensure end-users cannot bypass a validation by a webhook by removing resource label(s) which are used in the webhook's `objectSelector`.
  - Example: The [`ExtensionLabels` admission plugin](https://pkg.go.dev/github.com/gardener/gardener/plugin/pkg/global/extensionlabels#Register) maintains the `provider.extensions.gardener.cloud/<type>=true` label on CloudProfile, Seed, Shoot and other resources. The admission components of provider extensions use this label to filter for resources with the corresponding provider type (see [example usage](https://github.com/gardener/gardener-extension-provider-aws/blob/dd7dc94f049d7c8cf59f8211446f9015441d788c/pkg/admission/validator/webhook.go#L49-L51)). The `ExtensionLabels` admission plugin ensure the labels on `CREATE` and `UPDATE` requests. End-user cannot remove the `provider.extensions.gardener.cloud/<type>=true` label to bypass the provider extension admission webhook.

> - Use field-level validation errors according to their semantics.

Reviewer 3: ```suggestion
- Use [field-level validation errors](#field-level-validation-errors) according to their semantics.
```

  - Use `field.Required` for empty or null value; use `field.Invalid` for invalid value; use `field.Duplicate` for a duplicate value, etc. See [Field-level Validation Errors](#field-level-validation-errors).
- When introducing a new field, consider if it should be immutable or not.
  - Consider if updates to the field value should be allowed and are supported by the underlying controller. If not, consider making the field immutable. Add a doc string to the field to denote the immutability constraint.
- Introducing new validation for existing field or making existing validation more restrictive might be a breaking change.

>   - When working with existing field, aim to add validation which is obvious and unlikely to break a working functionality.

Reviewer 3: I think [ratcheting](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#ratcheting-validation) should be linked and mentioned in this context.

Author: Thanks for sharing!

  - The package also contains validation utilities for checking if Shoot networks intersect with Seed networks or if Shoot networks intersect between each other.
- [`github.com/gardener/gardener/pkg/utils/validation/kubernetes/core`](https://pkg.go.dev/github.com/gardener/gardener/pkg/utils/validation/kubernetes/core), aliased as `kubernetescorevalidation`
  - The package contains copy of several utility functions from [k8s.io/kubernetes/pkg/apis/core/validation](https://pkg.go.dev/k8s.io/kubernetes/pkg/apis/core/validation) and [k8s.io/kubernetes/pkg/apis/core/helper](https://pkg.go.dev/k8s.io/kubernetes/pkg/apis/core/helper). This is done to prevent importing the `k8s.io/kubernetes` dependency in Gardener.
- [`github.com/gardener/gardener/pkg/utils/validation`](https://pkg.go.dev/github.com/gardener/gardener/pkg/utils/validation), aliased as `validationutils`

>   - The package provides validation utilities for component configuration APIs.

Author: A note for me, check how to validate IP address.

cc @RadaBDimitrova

Author: I pushed f3625caf829f99ab31481de8f5e65114f11396e0.
